File: algorithms/python/advantageShuffle/advantageShuffle.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/advantage-shuffle/
# Author: Miao Zhang
# Date:   2021-03-19

class Solution:
    def advantageCount(self, A: List[int], B: List[int]) -> List[int]:
        A.sort()
        res = []
        for b in B:
            idx = bisect.bisect_right(A, b)
            if idx == len(A): idx = 0
            res.append(A[idx])
            A.pop(idx)
        return res

File: algorithms/cpp/advantageShuffle/advantageShuffle.cpp
// Source: https://leetcode.com/problems/advantage-shuffle/
// Author: Miao Zhang
// Date:   2021-03-19

class Solution {
public:
    vector<int> advantageCount(vector<int>& A, vector<int>& B) {
        multiset<int> s(begin(A), end(A));
        vector<int> res;
        for (int b: B) {
            auto it = s.upper_bound(b);
            if (it == s.end()) it = s.begin();
            res.push_back(*it);
            s.erase(it);
        }
        return res;
    }
};
